irepela/homework_11/mine_sweeper.py

```diff
@@ -0,0 +1,175 @@
+        """
+        Return cells, which are yet not opened and do not contain mines
+        """
+        return [(x, y)
+                for x in list(range(self.width))
+                for y in list(range(self.height))
+                if (x, y) not in self.mines and (x, y) not in self.guesses]
+
+    def in_field(self, coord: tuple) -> bool:
+        """
+        Check that coord is in field
+        """
+        return coord in self.empty_cells()
+
+    def is_mine(self, coord: tuple) -> bool:
+        """
+        Check if there is a mine under coord
+        """
+        return coord in self.mines
+
+    def mines_around(self, coord: tuple) -> int:
+        """
+        Calculate, how many mines are in neighbouring cells
+        """
+        x, y = coord
+        return sum((x + dx, y + dy) in self.mines
+                   for (dx, dy) in self.NEIGHBOURS)
+
+    def save_guess(self, coord: tuple):
+        """
+        Save user input
+        """
+        queue = [coord]
+        while queue:
+            x, y = queue.pop()
+            mines_around = self.mines_around((x, y))
+            self.guesses[(x, y)] = mines_around
+
+            # recursively open all neighbour cells if current cell == 0
+            if mines_around == 0:
+                neighborhood_cells = {(dx + x, dy + y) for dx, dy in self.NEIGHBOURS}
+                cells_to_check = neighborhood_cells.intersection(self.empty_cells())
+                queue.extend(cells_to_check)
+
+    def win(self):
+        """
+        Check if user wins the game
+        """
+        return not self.empty_cells()
+
+    def main(self):
+        while not self.win():
+            # show screen with current state
+            self.screen.update_screen()
+            # input from user
+            data = self.normalize_user_input()
+
+            # validate and normalize
+            # if error - catch, inform user and return to start
+            try:
+                self.validate_user_input(data)
+            except ValueError:
+                self.screen.show_alert(f"There should be only two int numbers x y: {data}")
+                continue
+            except IndexError:
+                self.screen.show_alert(f"Entered numbers: {data} are not in range: {width, height}")
+                continue
+
+            # if mine - end game
+            if self.is_mine(data):
+                self.screen.show_alert(f"There is a mine here, you have lost the game: {data}")
+                break
+
+            # save user input
+            self.save_guess(data)
+
+        # if win show win message
+        else:
+            self.screen.show_alert("You Won!!!")
+
+        # show screen last time
+        self.screen.update_screen(True)
+
+
+if __name__ == "__main__":
+    try:
+        width = int(input(Message.INPUT_ROWS_REQUEST).strip())
+        height = int(input(Message.INPUT_COLUMNS_REQUEST).strip())
+        mine_count = int(input(Message.INPUT_MINES_REQUEST).strip())
```

Comment:
create set of configs:
easy, medium, hard
and let user input 1 parameter: difficulty